kythe/go/extractors/validation/validation.go

```diff
@@ -0,0 +1,225 @@
+	Missing int
+	// A breakdown of the top missing files, by subdirectory.
+	TopMissing Coverage
+}
+
+// Coverage describes the missing subdirectories in a repo.
+type Coverage struct {
+	// The missing paths.
+	Paths []Dir
+}
+
+// The number of missing paths to print in output.
+const topMissingPathsLimit = 20
+const topMissingPathsDepth = 5
+
+// Dir describes a single missing subdirectory and how much is missing.
+type Dir struct {
+	// Path is a subdirectory from Repo root.
+	Path string
+	// Missing is the number of files in the Repo from the corresponding
+	// subdirectory that were missing from the Kzip.
+	Missing int
+}
+
+// Len is part of sort.Interface.
+func (c *Coverage) Len() int {
+	return len(c.Paths)
+}
+
+// Swap is part of sort.Interface.
+func (c *Coverage) Swap(i, j int) {
+	c.Paths[i], c.Paths[j] = c.Paths[j], c.Paths[i]
+}
+
+// Less is part of sort.Interface.
+func (c *Coverage) Less(i, j int) bool {
+	return c.Paths[i].Missing < c.Paths[j].Missing
+}
+
+// Validate opens a kzip file and compares all of its contained files with those
+// present in the provided repo path.  It returns results explaining what
+// percentage of files are covered in the kzip.
+func (h Harness) Validate() (*Result, error) {
+	log.Printf("Gathering validation data for %s", h.Kzip)
+	fromKzip, err := h.filenamesFromKzip(h.Kzip)
+	if err != nil {
+		return nil, fmt.Errorf("reading from kzip %s: %v", h.Kzip, err)
+	}
+	fromRepo, err := h.filenamesFromPath(h.Repo)
+	if err != nil {
+		return nil, fmt.Errorf("reading from repo %s: %v", h.Repo, err)
+	}
+	missingFromKzip := map[string]bool{}
+	for f := range fromRepo {
+		if _, ok := fromKzip[f]; !ok {
+			missingFromKzip[f] = true
+		}
+	}
+
+	// Try to calculate the most missing subdirectories.
+	topMissing := Coverage{}
+	// Only try to go so many paths deep (src/java/com/domain/sub).
+	// TODO(danielmoy): be more intelligent about this.
+	for i := 1; i <= topMissingPathsDepth; i++ {
+		topMissing = getTopMissing(missingFromKzip, topMissingPathsLimit, i)
+		if len(topMissing.Paths) >= topMissingPathsLimit {
+			break
+		}
+	}
+
+	if h.MissingOutput != nil {
+		err = printMissing(missingFromKzip, *h.MissingOutput)
+	}
+
+	return &Result{
+		FilesInKzip: len(fromKzip),
+		FilesInRepo: len(fromRepo),
+		Missing:     len(missingFromKzip),
+		TopMissing:  topMissing,
+	}, err
+}
+
+func (h Harness) filenamesFromKzip(path string) (map[string]bool, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("opening kzip %s: %v", path, err)
+	}
+	fi, err := f.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("getting kzip size %s: %v", path, err)
+	}
+	r, err := kzip.NewReader(f, fi.Size())
+	if err != nil {
+		return nil, fmt.Errorf("opening kzip reader %s: %v", path, err)
+	}
+
+	ret := map[string]bool{}
+
+	if err := r.Scan(func(cu *kzip.Unit) error {
+		if cu.Proto != nil {
+			for _, f := range cu.Proto.SourceFile {
+				ret[f] = true
```

Comment:
Does it matter that we aren't checking the extension here in the kzip, but we are when we examine the file system?